File: client-scalaz/src/main/scala/org/openapitools/client/api/OrderItemRequest.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import OrderItemRequest._

case class OrderItemRequest (
  orderItemType: Option[OrderItemType],
orderItemId: Option[Long],
offerName: Option[String],
orderCustomType: Option[String],
orderCustomId: Option[String],
retailerLocationId: Option[Long],
quantity: Option[Integer],
amount: Option[Double],
tax: Option[Double],
reserveStartDate: Option[Long],
reserveEndDate: Option[Long],
message: Option[String])

object OrderItemRequest {
  import DateTimeCodecs._
  sealed trait OrderItemType
  case object OFFER extends OrderItemType
  case object CUSTOM extends OrderItemType

  object OrderItemType {
    def toOrderItemType(s: String): Option[OrderItemType] = s match {
      case "OFFER" => Some(OFFER)
      case "CUSTOM" => Some(CUSTOM)
      case _ => None
    }

    def fromOrderItemType(x: OrderItemType): String = x match {
      case OFFER => "OFFER"
      case CUSTOM => "CUSTOM"
    }
  }

  implicit val OrderItemTypeEnumEncoder: EncodeJson[OrderItemType] =
    EncodeJson[OrderItemType](is => StringEncodeJson(OrderItemType.fromOrderItemType(is)))

  implicit val OrderItemTypeEnumDecoder: DecodeJson[OrderItemType] =
    DecodeJson.optionDecoder[OrderItemType](n => n.string.flatMap(jStr => OrderItemType.toOrderItemType(jStr)), "OrderItemType failed to de-serialize")

  implicit val OrderItemRequestCodecJson: CodecJson[OrderItemRequest] = CodecJson.derive[OrderItemRequest]
  implicit val OrderItemRequestDecoder: EntityDecoder[OrderItemRequest] = jsonOf[OrderItemRequest]
  implicit val OrderItemRequestEncoder: EntityEncoder[OrderItemRequest] = jsonEncoderOf[OrderItemRequest]
}

File: client-clojure/src/sirqul_io_t_platform/specs/order_item_request.clj
(ns sirqul-io-t-platform.specs.order-item-request
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            )
  (:import (java.io File)))


(def order-item-request-data
  {
   (ds/opt :orderItemType) string?
   (ds/opt :orderItemId) int?
   (ds/opt :offerName) string?
   (ds/opt :orderCustomType) string?
   (ds/opt :orderCustomId) string?
   (ds/opt :retailerLocationId) int?
   (ds/opt :quantity) int?
   (ds/opt :amount) float?
   (ds/opt :tax) float?
   (ds/opt :reserveStartDate) int?
   (ds/opt :reserveEndDate) int?
   (ds/opt :message) string?
   })

(def order-item-request-spec
  (ds/spec
    {:name ::order-item-request
     :spec order-item-request-data}))
